Conch/lib/Conch/Orc/Lifecycle.pm

```diff
@@ -0,0 +1,472 @@
+		}
+	}
+
+	my $ret;
+	try {
+		if($self->id) {
+			$ret = $db->update(
+				'workflow_lifecycle',
+				\%fields,
+				{ id => $self->id },
+				{ returning => [qw(id created updated)] }
+			)->hash;
+		} else {
+			$ret = $db->insert(
+				'workflow_lifecycle',
+				\%fields,
+				{ returning => [qw(id created updated)] }
+			)->hash;
+		}
+	} catch {
+		Mojo::Exception->throw(__PACKAGE__."->save: $_");
+		return undef;
+	};
+
+	$ret = _fixup_timestamptzs($ret);
+
+	$self->_set_id($ret->{id});
+	$self->_set_created(Conch::Time->new($ret->{created}));
+	$self->_set_updated(Conch::Time->new($ret->{updated}));
+		
+	if($ret->{deactivated}) {
+		$self->deactivated($ret->{deactivated});
+	}
+
+	return $self;
+}
+
+=head2 append_workflow
+
+	$lifecycle->append_workflow($workflow->id);
+
+Append a workflow UUID to the lifecycle's plan. Does nothing if the lifecycle
+has not been saved (lacks an ID).
+
+A unique constraint exists on the lifecycle id and workflow id, meaning that a
+workflow can only exist in a lifecycle's plan a single time. If an
+C<append_workflow> call attempts to add a duplicate workflow id to the
+plan, the plan will remain unchanged.
+
+=cut
+
+sub append_workflow ($self, $workflow_uuid) {
+	return $self unless $self->id;
+
+	my @plan = $self->plan->@*;
+	push @plan, $workflow_uuid;
+	$self->_set_plan(\@plan);
+
+	return $self->_rebuild_plan;
+}
+
+=head2 add_workflow
+
+	$lifecycle->add_workflow($workflow->id, 2);
+
+Add a workflow UUID to the lifecycle's plan at the position indicated. The
+position numeral is an array index and starts counting at 0.
+
+A unique constraint exists on the lifecycle id and workflow id, meaning that a
+workflow can only exist in a lifecycle's plan a single time. If an
+C<add_workflow> call attempts to add a duplicate workflow id to the
+plan, the plan will remain unchanged.
+
+=cut
+
+sub add_workflow ($self, $workflow_uuid, $order) {
+	my @p = $self->plan->@*;
+	my @new_plan;
+	foreach my $idx (0 .. $#p) {
+		if ($idx == $order) {
+			push @new_plan, $workflow_uuid;
+		}
+		push @new_plan, $p[$idx];
+	}
+
+	$self->_set_plan(\@new_plan);
+
+	return $self->_rebuild_plan;
+}
+
+=head2 remove_workflow
+
+	$lifecycle->remove_workflow($workflow->id);
+
+Removes all occurences of a workflow UUID in the lifecycle's plan. Does nothing
+if the lifecycle has not been saved (lacks an ID).
+
+=cut
+
+sub remove_workflow ($self, $workflow_uuid) {
+	return $self unless $self->id;
+
+	my @p = $self->plan->@*;
+	my @new_plan;
+	foreach my $idx (0 .. $#p) {
+		unless ($p[$idx] eq $workflow_uuid) {
+			push @new_plan, $p[$idx];
+		}
+	}
+	$self->_set_plan(\@new_plan);
+
+	return $self->_rebuild_plan;
+}
+
+
+# Blows away the existing join records and recreates them using the array 
+# indices on ->plan as the plan_order field
+sub _rebuild_plan ($self) {
+	return $self unless $self->id;
+	my $db = Conch::Pg->new->db;
+	my $tx = $db->begin;
+
+	my @p = uniq($self->plan->@*);
```

Comment:
If there's duplicates in the plan, this will drop the latter duplicates. This may have surprising results. If I add workflow ID 1 as order 2 but workflow ID 1 is already a member of the plan as order 4, the order 4 plan will be silently dropped.

This is only an issue for adding new workflows. You may want to check if a workflow already exists in the add_workflow controller and return 409 if so. To reorder workflows, the user or client should explicitly remove and re-add the workflow at the desired location.

Reply:
The DB layer should be fixed for this use case now. The plan modification routines now check if the workflow is or isn't in the list and don't make a trip to the database if things are not as expected. It shouldn't be possible, with those changes, to get duplicate workflows into the plan.

Reply:
@lseppala this comment :)

Reply:
Storing duplicates wasn't the issue. Sorry for the confusion. More, if you had two duplicates in `$self->plan` before calling `_rebuild_plan` (as is the case if an already-stored workflow is added in a request), it may silently alter or ignore the ordering. 

This is resolved by the checks you added in add_workflow and append_workflow. This means re-adding a previously stored workspace will always be ignored, even if the new request has a different order. One might consider returning an error in the controller if the workflow, but it's fine as it's consistent.